Check for overflow on left shift of signed ints

tautschnig opened the pull request:
This is undefined behaviour according to 6.5.7 paragraph 4 in C11.

tautschnig committed:
Check for overflow on left shift of signed ints

This is undefined behaviour according to 6.5.7 paragraph 4 in C11.

kroening commented:
BTW, clang also warns:
warning: shifting a negative signed value is undefined
What does the standard say?

kroening commented on src/solvers/flattening/boolbv_overflow.cpp:
```diff
@@ -96,6 +95,43 @@ literalt boolbvt::convert_overflow(const exprt &expr)
+        ID_gt,
+        bv_utils.build_constant(old_size, bv1.size()),
+        bv_utilst::representationt::UNSIGNED);
+
```
I believe that the right-hand-side operand may be either signed or unsigned. I think the above will only work if the right-hand side has enough bits (i.e., sign extension may be needed).

tautschnig replied:
I was thinking that negative shifts should not be caught here, but indeed I need to make sure that this case is truly ignored.

tautschnig replied:
I've added this check in the updated revision.

kroening commented on src/solvers/flattening/boolbv_overflow.cpp:
```diff
@@ -96,6 +95,43 @@ literalt boolbvt::convert_overflow(const exprt &expr)
+      // get top result bits plus sign bit
+      DATA_INVARIANT(old_size!=0, "zero-size operand");
+      result.erase(result.begin(), result.begin()+old_size-1);
+    }
```
Did you intend that left-shifts of negative numbers are ruled out (see other comment)?

tautschnig commented:
> BTW, clang also warns:
warning: shifting a negative signed value is undefined
What does the standard say?

Where is it that this warning appears?

kroening commented:
Say try
printf("%d\n", -1<<((sizeof(int)*8-1)));

kroening commented on regression/cbmc/Overflow_Subtraction1/main.c:
```diff
@@ -1,4 +1,4 @@
-#define INT_MIN (-1<<(sizeof(int)*8-1))
+#define INT_MIN (1U<<(sizeof(int)*8-1))
 
```
This is now at least an unusual definition of INT_MIN

tautschnig replied:
Yet it has defined behaviour. Some discussion can also be found at https://blog.regehr.org/archives/738. Actual header files appear to be using `-INT_MAX-1`.

tautschnig replied:
Reverted/fixed in the latest revision.

kroening commented:
The following seems to be the right approach on shifting negative numbers to the left (do not confuse with negative shift distances): It's an overflow iff a) the new sign is 0, or b) any of the bits that are shifted out to the left are 0.

tautschnig commented:
> The following seems to be the right approach on shifting negative numbers to the left (do not confuse with negative shift distances): It's an overflow iff a) the new sign is 0, or b) any of the bits that are shifted out to the left are 0.

Thank you, you are of course right. This has now been addressed.

tautschnig commented:
> printf("%d\n", -1<<((sizeof(int)*8-1)));

It seems actually all shifts of negative numbers are undefined:
"If E1 has a signed type and nonnegative value, and E1 × 2^E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined."

Maybe the question is: should negative numbers be flagged as overflows or as invalid shifts?

peterschrammel commented:
I think shifts of negative numbers should be invalid shifts.

tautschnig committed:
Shifts of non-integers and left shifts of negative integers are undefined

6.5.7 of C11: "Each of the operands shall have integer type."

Paragraph 4: "If E1 has a signed type and nonnegative value, and E1 x 2^E2 is
representable in the result type, then that is the resulting value; otherwise,
the behavior is undefined."

kroening commented:
I would also go for "invalid shift"

tautschnig commented:
@peterschrammel @kroening Done in 1e3712c.
